.github/copilot-instructions.md

```diff
@@ -0,0 +1,837 @@
+    THANDLE_ASSIGN(MY_OBJECT)(&local_handle, NULL);  // Decrements ref count, frees if count reaches 0
+    
+    // Or let caller manage cleanup by not setting to NULL
+}
+```
+
+### Advanced THANDLE Patterns
+
+#### Function Parameters and Return Values
+```c
+// Passing THANDLE by value (preferred pattern)
+int process_object(THANDLE(MY_OBJECT) object)
+{
+    if (object == NULL)
+    {
+        return -1;
+    }
+    
+    // Use object->data directly
+    return object->data;
+    // No cleanup needed - caller retains ownership
+}
+
+// Returning THANDLE (caller gets ownership)
+THANDLE(MY_OBJECT) create_object_with_data(int initial_data)
+{
+    THANDLE(MY_OBJECT) result = THANDLE_MALLOC(MY_OBJECT)(dispose_function);
+    if (result != NULL)
+    {
+        result->data = initial_data;
+    }
+    return result;  // Caller gets ownership
+}
+
+// Output parameter pattern (for optional results)
+int try_get_object(THANDLE(MY_OBJECT)* out_object)
+{
+    THANDLE(MY_OBJECT) temp = create_some_object();
+    if (temp == NULL)
+    {
+        return -1;
+    }
+    
+    THANDLE_MOVE(MY_OBJECT)(out_object, &temp);  // Transfer to caller
+    return 0;
+}
+```
+
+#### Collections and Data Structures
+```c
+// THANDLE in arrays
+typedef struct OBJECT_COLLECTION_TAG
+{
+    THANDLE(MY_OBJECT)* objects;
+    size_t count;
+    size_t capacity;
+} OBJECT_COLLECTION;
+
+void add_object_to_collection(OBJECT_COLLECTION* collection, THANDLE(MY_OBJECT) object)
+{
+    // Resize if needed
+    if (collection->count >= collection->capacity)
+    {
+        // ... resize logic ...
+    }
+    
+    // Store shared reference
+    THANDLE_ASSIGN(MY_OBJECT)(&collection->objects[collection->count], object);
+    collection->count++;
+}
+
+void cleanup_collection(OBJECT_COLLECTION* collection)
+{
+    // Clean up all handles
+    for (size_t i = 0; i < collection->count; i++)
+    {
+        THANDLE_ASSIGN(MY_OBJECT)(&collection->objects[i], NULL);  // Decrements ref count
+    }
+    free(collection->objects);
+    collection->objects = NULL;
+    collection->count = 0;
+}
+```
+
+#### Thread Safety and Shared Objects
+```c
+// THANDLE provides thread-safe reference counting
+// Multiple threads can safely share THANDLE instances
+
+typedef struct SHARED_CONTEXT_TAG
+{
+    THANDLE(MY_OBJECT) shared_object;
+    // ... other fields
+} SHARED_CONTEXT;
+
+// Worker thread function
+DWORD WINAPI worker_thread(LPVOID param)
+{
+    SHARED_CONTEXT* context = (SHARED_CONTEXT*)param;
+    
+    // Create local reference to shared object
+    THANDLE(MY_OBJECT) local_ref;
+    THANDLE_ASSIGN(MY_OBJECT)(&local_ref, context->shared_object);
+    
+    // Use object safely - it won't be freed while this thread holds a reference
+    if (local_ref != NULL)
+    {
+        // ... work with local_ref->data ...
+    }
+    
+    // Manual cleanup required before thread exits
+    THANDLE_ASSIGN(MY_OBJECT)(&local_ref, NULL);
```

Comment:
this is all very wrong. the THANDLE has been assigned (read: inc_ref'd) by the thread creator way before the thread is created  and code starts to execute. Usually it sits in the thread `param`.

Then, when the thread is done "looking at it" (read: releases ownership) it will assign NULL to something.

Reply:
Asked it to rewrite it. And rewritten it is.